fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/ComparatorCursorBase.java

```diff
@@ -0,0 +1,243 @@
+    @Nonnull
+    private static final Set<StoreTimer.Event> duringEvents = Collections.singleton(FDBStoreTimer.Events.QUERY_COMPARATOR);
+    @Nonnull
+    private static final Set<StoreTimer.Count> matchesCounts = Collections.singleton(FDBStoreTimer.Counts.QUERY_COMPARATOR_MATCH);
+    @Nonnull
+    private static final Set<StoreTimer.Count> mismatchesCounts = Collections.singleton(FDBStoreTimer.Counts.QUERY_COMPARATOR_MISMATCH);
+    @Nonnull
+    private static final Set<StoreTimer.Count> compareCounts = Collections.singleton(FDBStoreTimer.Counts.QUERY_COMPARATOR_COMPARED);
+
+    // The "reference plan" whose values are to be compared with the rest of the cursors, and whose values are actually returned
+    private final int referencePlanIndex;
+
+    private boolean errorLogged = false;
+
+    @Nonnull
+    private final Supplier<String> planStringSupplier;
+    @Nonnull
+    private final Supplier<Integer> planHashSupplier;
+
+    protected ComparatorCursorBase(@Nonnull List<KeyedMergeCursorState<T>> cursorStates,
+                                   @Nullable FDBStoreTimer timer,
+                                   final int referencePlanIndex,
+                                   @Nonnull final Supplier<String> planStringSupplier,
+                                   @Nonnull final Supplier<Integer> planHashSupplier) {
+        super(cursorStates, timer);
+        this.referencePlanIndex = referencePlanIndex;
+        this.planStringSupplier = planStringSupplier;
+        this.planHashSupplier = planHashSupplier;
+    }
+
+    /**
+     * Compute the set of next result states. This would wait for the states to have a value, then compare all values
+     * to make sure they are identical.
+     * If one of the values doesn't match the reference, or one of the cursors ends prematurely, the cursor logs the
+     * failure.
+     * Return empty list in case one of the states is done (but not exhausted), signaling that subsequent iteration
+     * using a continuation is required.
+     *
+     * @return the list of states included in the next result
+     */
+    @Override
+    @Nonnull
+    protected CompletableFuture<List<KeyedMergeCursorState<T>>> computeNextResultStates() {
+        final List<KeyedMergeCursorState<T>> cursorStates = getCursorStates();
+        // Wait for all cursors to have a valid state
+        return whenAll(cursorStates).thenApply(vignore -> {
+            // if all cursors have a value, compare and continue
+            if (cursorStates.stream().allMatch(cursorState -> hasNext(cursorState))) {
+                compareAllStates(cursorStates);
+                // Regardless of comparison result, return all states (comparison logged results)
+                return cursorStates;
+            }
+            // some cursors have no next result, if all are exhausted, we are done
+            long exhaustedCount = cursorStates.stream().filter(cursorState -> isSourceExhausted(cursorState)).count();
```

Comment:
I may be missing something, but the intent was to allow cursors to "independently" progress (using whatever pace they have) and to make sure that when they all return a result, that result gets compared. For that, we should allow cursors to hit their scan limit and continue later (using a continuation), and the only "real" stopage is the cursor reaching exhaustion.

Reply:
As part of the logic change for the out-of-band limits, added more logic about in-band limit as well. Seems like if we have any combination of exhausted plans and in-band-limit hit then we have a "early termination error" (since the reference plan and at least one other plan have to be different). Please take a look